pkg/retry/azure_error.go

```diff
@@ -327,47 +327,85 @@ func HasStatusForbiddenOrIgnoredError(err error) bool {
 	return false
 }
 
-// ParseRawError parse the error message in the rawError and unmarshal it into RawErrorContainer
-func ParseRawError(rawError string) (*RawErrorContainer, error) {
-	reg := regexp.MustCompile(`^(?:[^{]*)([\s\S]*)$`)
-	matches := reg.FindStringSubmatch(rawError)
-	if len(matches) != 2 {
-		klog.V(4).Infof("skipping parsing because the format of the raw error message %q is not the expected one")
-		return nil, nil
+// GetVMSSMetadataByRawError gets the vmss name by parsing the error message
+func GetVMSSMetadataByRawError(err *Error) (string, string, error) {
+	if err == nil || !isErrorLoadBalancerInUseByVirtualMachineScaleSet(err.RawError.Error()) {
+		return "", "", nil
 	}
 
-	rawErrorMap := make(map[string]*RawErrorContainer)
-	err := json.Unmarshal([]byte(matches[1]), &rawErrorMap)
-	if err != nil {
-		return nil, err
+	reg := regexp.MustCompile(`.*/subscriptions/(?:.*)/resourceGroups/(.*)/providers/Microsoft.Compute/virtualMachineScaleSets/(.+).`)
+	matches := reg.FindStringSubmatch(err.ServiceErrorMessage())
+	if len(matches) != 3 {
+		return "", "", fmt.Errorf("GetVMSSMetadataByRawError: couldn't find a VMSS resource Id from error message %s", err.RawError)
 	}
 
-	return rawErrorMap["error"], nil
+	return matches[1], matches[2], nil
 }
 
-// IsErrorLoadBalancerInUseByVirtualMachineScaleSet determines if the Error is
+// isErrorLoadBalancerInUseByVirtualMachineScaleSet determines if the Error is
 // LoadBalancerInUseByVirtualMachineScaleSet
-func IsErrorLoadBalancerInUseByVirtualMachineScaleSet(rawError string) bool {
+func isErrorLoadBalancerInUseByVirtualMachineScaleSet(rawError string) bool {
 	return strings.Contains(rawError, "LoadBalancerInUseByVirtualMachineScaleSet")
 }
 
-// GetVMSSMetadataByRawError gets the vmss name by parsing the error message
-func GetVMSSMetadataByRawError(rawError string) (string, string, error) {
-	if !IsErrorLoadBalancerInUseByVirtualMachineScaleSet(rawError) {
-		return "", "", nil
+const (
+	// OperationNotAllowed is an umbrella errrfor a lot of errors
+	OperationNotAllowed string = "OperationNotAllowed"
+	// QuotaExceeded falls under OperationNotAllowed error code but we make it more specific here
+	QuotaExceeded string = "QuotaExceeded"
+)
+
+// ServiceRawError wraps the RawError field satisfying autorest.ServiceError
```

Comment:
RawError under this package will always embed either azure.ServiceError or azure.DetailedError.

We don't care about azure.DetailedError for now since they seem to arise from ARM-context outside the providers.